main.go

```diff
@@ -14,27 +21,81 @@ func main() {
 		log.Fatal(err)
 	}
 
+	if err = run(config); err != nil {
+		config.Logger.Error(errors.Wrap(err, "startup"))
+		os.Exit(1)
+	}
+}
+
+const shutdownTimeoutSeconds = 3
+
+func run(conf *sat.Config) error {
+	localLogger := conf.Logger.CreateScoped("main.run")
+
+	traceProvider, err := sat.SetupTracing(conf.TracerConfig, conf.Logger)
+	if err != nil {
+		return errors.Wrap(err, "setup tracing")
+	}
+	defer traceProvider.Shutdown(context.Background())
+
 	// initialize Reactr, Vektor, and Grav and wrap them in a sat instance
-	s, err := sat.New(config)
+	s, err := sat.New(conf, traceProvider)
 	if err != nil {
-		log.Fatal(err)
+		return errors.Wrap(err, "sat.New")
 	}
 
-	if config.UseStdin {
-		if err := s.ExecFromStdin(); err != nil {
-			log.Fatal(err)
+	if conf.UseStdin {
```

Comment:
We should probably do this check before the tracer is set up, because there's no reason to set up the tracer if we're just going to execute the stdin job and quit

Reply:
That will create a circular dependency though. If we use stdin, we still need to exec on sat, which means we need to create a new sat, which means pass a tracer to sat, so we can't execute without a tracer.

An empty tracer is possible though, I don't think that's going to be bad, especially given when it runs locally, it doesn't need to scale much, so the overhead of setting up an empty tracer is negligible.

Reply:
That adds complexity and logic into `sat.New`. The point of initializing the tracer before calling `sat.New` is to move decisions and logic that doesn't need to be within `sat.New` out of it. Plus it returns a closer which  we'd need to defer, and if we have two different ways of handling tracing, that defeats the purpose of simplifying startup / shutdown.

Alternatively we can make a decision in `main.go` based on the config, and call a different `run` that skips much of the initialization that we don't need if it's in stdin mode. Let me put up a PR to this for that.

Reply:
Added a PR that would be merged into this branch here: https://github.com/suborbital/sat/pull/58